File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    let left = 1;
    let right = diffs.reduce((acc,cur) => Math.max(acc, cur), 1);
    
    function calcDiff(mid) {
        let result = 0;
        
        for (let index = 0; index < diffs.length; index++) {
            if (mid < diffs[index]) {
                const mistakeCount = diffs[index] - mid;
                const prevTime = index > 0 ? times[index-1] : 0;
                result += mistakeCount * (times[index] + prevTime);
            }
            result += times[index];
            if (result > limit) {
                return false;
            }
        }
        
        return result <= limit;
    }
    
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        
        if (calcDiff(mid)) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    
    return left;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    def solve_puzzle(level):
        time = 0
        
        for idx in range(len(diffs)):
            if diffs[idx] > level:
                wrong_count = diffs[idx] - level
                time_prev = 0 if idx == 0 else times[idx-1]
                time += wrong_count * (times[idx] + time_prev)
            time += times[idx]
            
            if time > limit:
                return False
        
        return time <= limit
    
    left, right = 1, max(diffs)
    
    while left <= right:
        mid = (left + right) // 2
        
        if solve_puzzle(mid):
            right = mid - 1
        else:
            left = mid + 1
    
    return left
            